storage2/src/storage.rs

```diff
@@ -29,16 +64,129 @@ impl Storage {
+                span.in_scope(|| {
+                    let snap = self.latest_snapshot.read().clone();
+                    let jmt = JellyfishMerkleTree::new(snap.as_ref());
+                    // Write the unwritten changes from the state to the JMT.
+                    let (jmt_root_hash, batch) = jmt.put_value_set(
+                        state
+                            .unwritten_changes
+                            .into_iter()
+                            .map(|x| (KeyHash::from(x.0), x.1))
+                            .collect(),
+                        new_version,
+                    )?;
+
+                    // Apply the JMT changes to the DB.
+                    self.write_node_batch(&batch.node_batch)?;
+                    tracing::trace!(?jmt_root_hash, "wrote node batch to backing store");
+
+                    // TODO: 3. write the index tables to RocksDB
+                    // not yet implemented because the API for vectorized keys is not yet designed
+
+                    // 4. update the snapshot
+                    // Now that we've successfully written the new nodes, update the version.
+                    let jmt_version = new_version;
+                    let snapshot = db.snapshot();
+                    // Obtain the write-lock for the latest snapshot, and replace it with the new snapshot.
+                    let mut guard = self.latest_snapshot.write();
+                    *guard = Arc::new(Snapshot::new(snapshot, jmt_version, db));
+                    // Drop the write-lock (this will happen implicitly anyways, but it's good to be explicit).
+                    drop(guard);
+                    anyhow::Result::<()>::Ok(())
+                })
+            })
+            .unwrap()
+            .await
+            .unwrap();
+
+        Ok(())
     }
 }
+
+impl TreeWriter for Storage {
+    /// Writes a node batch into storage.
+    //TODO: Change JMT traits to accept owned NodeBatch
+    fn write_node_batch(&self, node_batch: &NodeBatch) -> Result<()> {
+        let db = self.db;
+        let node_batch = node_batch.clone();
+
+        // The writes have to happen on a separate spawn_blocking task, but we
+        // want tracing events to occur in the context of the current span, so
+        // propagate it explicitly:
+        let span = Span::current();
+
+        for (node_key, node) in node_batch.clone() {
+            let key_bytes = &node_key.encode()?;
+            let value_bytes = &node.encode()?;
+            tracing::trace!(?key_bytes, value_bytes = ?hex::encode(&value_bytes));
+
+            let jmt_cf = db.cf_handle("jmt").expect("jmt column family not found");
+            db.put_cf(jmt_cf, key_bytes, &value_bytes)?;
+        }
+
+        Ok(())
+    }
+}
+
+// TODO: maybe these should live elsewhere?
+fn get_rightmost_leaf(db: &DB) -> Result<Option<(NodeKey, LeafNode)>> {
+    let span = Span::current();
+
+    // TODO: should this be in a tokio task like before?
```

Comment:
This should not be in a tokio task, because `get_rightmost_leaf` is not an async function.

Reply:
But whatever calls it should be in a Tokio task because it is calling out to the blocking rocksdb code?

Reply:
At the outermost point possible, it should be in a `spawn_blocking` call, yes. But in order to make things most efficient, we should push that boundary as far out as possible, because each `spawn_blocking` corresponds to a thread allocated on the blocking thread pool.

Reply:
Though I guess this storage code will ultimately be abstracted away behind another layer of API and the higher-level `pd` code shouldn't have to worry about spawning threads when DB access is performed.

Reply:
That's right: the storage API exposed to `pd` will be an `async` API, but we want to only spawn one thread for each call to a function on `Storage`, rather than many. The `Storage` interface will end up wrapping each method's body in a `spawn_blocking`, I think.

Reply:
In order to abstract away the storage code, though, we'll need to implement the trait differently: https://github.com/penumbra-zone/penumbra/pull/1523/files#r995878288

Otherwise, it will leak into the public API.